core/application/chain_spec.hpp

```diff
@@ -7,13 +7,15 @@
 #define KAGOME_CHAIN_SPEC_HPP
 
 #include <libp2p/peer/peer_info.hpp>
-#include "application/genesis_raw_data.hpp"
 #include "crypto/ed25519_types.hpp"
 #include "crypto/sr25519_types.hpp"
 #include "primitives/block.hpp"
 
 namespace kagome::application {
 
+  using CodeSubstitutes = std::map<primitives::BlockHash, common::Buffer>;
```

Comment:
The new code may arrive in several blocks, since we have a competitive block production (there may be several leaders in the slot). Therefore, there may be several copies of the same code in memory. I think it would be good to have additional map layer: map of block hashes to hashes of code, and map of hashes of code to shared pointers to buffer cintaining code.

Reply:
I agree. But it has nothing to do with code substitutes. CodeSubstitutes resembles chainspec json structure.

Reply:
runtime::Module is the structure that stores parsed code, thus it is better to have map<Hash<Code>, Module> rather than map<Hash<State>, Module>. Wonder how productive to check code hash on every block. Actually substrate uses such an approach.

Reply:
Hm... Need to think about it. If if difficult to implement it in this PR, open the issue for that

Reply:
Didn't get the point really. What wrong with several copies in competitive block production? We may discuss it on status.